### PR TITLE
Rss Bar and Legend Link (Job List Footer) Added Classes and IDs to enable easy styling for external themes

### PR DESCRIPTION
**Pull request by clouless**
**TL;DR: Minor improvement to give RSS Bar Classes and IDs to enable easy styling.**

image

### Proposed changelog entries

* RSS bar now has IDs and Classes on all its elements, so that external themes can better style the rss bar (which currently is almost not possible).
  * RSS bar Icons now obey the convention for icon classes `icon-NAME icon-SIZE` e.g. `icon-rss icon-sm`
  * Migrated inline styles to styling by classes/ids. Migrated HTML styling attributes like `width` to CSS.

### Submitter checklist

- ~[ ] JIRA issue is well described~ 
  * *Minor improvement, so no JIRA issue.*
- [x] Changelog entry appropriate for the audience affected by the change:
  * Internal, developer
- [x] Appropriate autotests or explanation to why this change has no tests
  * I has not tests, since changes a minor UI changes that would require an E2E test.
  * Changes have been tested in multiple browsers.
- ~[ ] For dependency updates: links to external changelogs and, if possible, full diffs~

### Desired reviewers

@i386 Maybe you remember me from the [PR on the favorite plugin](https://github.com/jenkinsci/favorite-plugin/pull/9/files). I would like to do the same to the RSS Bar and just give it proper classes and ids :)
If it is ok I would add some more small PRs on this topic for other elements that have missing classes/ids that make them hard to style.

### CODE REVIEW

**Comment by clouless** [comment]
fyi: Checks have failed because of minor GitHub downtime. How can we rerun the checks?

**Comment by daniel-beck** [comment]
@clouless Done

File: core/src/main/resources/lib/hudson/rssBar.jelly [hunk 1]
```diff
@@ -24,22 +24,33 @@ THE SOFTWARE.
...
+      border: 0;
+    }
+    .rss-bar-item {
+      padding-left: 1em;
+    }
+    #rss-bar {
+      margin:1em;
+    }
+  </style>
+  <div id="rss-bar" align="right">
```
Comment (recena) [inline core/src/main/resources/lib/hudson/rssBar.jelly]: 🐛 `align` attribute is not allowed on `div` element. Moreover, you would be mixing semantic and representation.

File: core/src/main/resources/lib/hudson/rssBar.jelly [hunk 2]
```diff
@@ -24,22 +24,33 @@ THE SOFTWARE.
 
 <?jelly escape-by-default='true'?>
 <j:jelly xmlns:j="jelly:core" xmlns:x="jelly:xml" xmlns:st="jelly:stapler" xmlns:d="jelly:define" xmlns:l="/lib/layout" xmlns:t="/lib/hudson" xmlns:f="/lib/form">
-  <div align="right" style="margin:1em">
-    <a href="${rootURL}/legend">${%Legend}</a>
-    <span style="padding-left:1em">
-      <a href="rssAll"><img src="${imagesURL}/atom.gif" border="0" alt="Feed" height="16" width="16"/></a>
+  <style type="text/css">
```
Comment (recena) [inline core/src/main/resources/lib/hudson/rssBar.jelly]: How not move this visual representation to `style.css`?

Comment (clouless) [inline core/src/main/resources/lib/hudson/rssBar.jelly]: Ok if you mean this style.css https://github.com/clouless/jenkins/blob/ui-classes-joblist-rss-links-and-legend/war/src/main/webapp/css/style.css

I can move it there

File: core/src/main/resources/lib/hudson/rssBar.jelly (hunk 1 above)
Comment (clouless) [inline core/src/main/resources/lib/hudson/rssBar.jelly]: @recena fixed.

File: core/src/main/resources/lib/hudson/rssBar.jelly [hunk 3]
```diff
@@ -24,22 +24,22 @@ THE SOFTWARE.
 
 <?jelly escape-by-default='true'?>
 <j:jelly xmlns:j="jelly:core" xmlns:x="jelly:xml" xmlns:st="jelly:stapler" xmlns:d="jelly:define" xmlns:l="/lib/layout" xmlns:t="/lib/hudson" xmlns:f="/lib/form">
-  <div align="right" style="margin:1em">
-    <a href="${rootURL}/legend">${%Legend}</a>
-    <span style="padding-left:1em">
-      <a href="rssAll"><img src="${imagesURL}/atom.gif" border="0" alt="Feed" height="16" width="16"/></a>
+  <div id="rss-bar">
```
Comment (recena) [inline core/src/main/resources/lib/hudson/rssBar.jelly]: IMHO, with this `id`would be enough. You don't need to fill the HTML markup with IDs

File: war/src/main/webapp/css/style.css [hunk 4]
```diff
@@ -1913,3 +1913,18 @@ body.no-sticker #bottom-sticker {
...
+}
+
+.rss-bar-item {
+    padding-left: 1em;
+}
+
+#rss-bar {
+    margin:1em;
+    text-align:right;
+}
```
Comment (recena) [inline war/src/main/webapp/css/style.css]:
```
/* rss-bar */

#rss-bar {
    margin: 1em;
    text-align: right;
}

#rss-bar .icon-rss {
    border: 0;
}

#rss-bar .rss-bar-item {
    padding-left: 1em;
}
```

More robust.

File: war/src/main/webapp/css/style.css [hunk 5]
```diff
@@ -1,18 +1,18 @@
 /*
  * The MIT License
- * 
+ *
```
Comment (recena) [inline war/src/main/webapp/css/style.css]: Someone is going to say "_Unrelated changes..._"

Comment (clouless) [inline war/src/main/webapp/css/style.css]: yeah sorry, the editor strips spaces at the line end. they shouldn't be there anyway, so you may see it as improvement :D

File: core/src/main/resources/lib/hudson/rssBar.jelly (hunk 3 above)
Comment (clouless) [inline core/src/main/resources/lib/hudson/rssBar.jelly]: the id thing seems a little to much, but I am developing a theme and there it is nice to have ids to write robust css. So I propose them to stay, since they are not hurting anyone.

File: war/src/main/webapp/css/style.css (hunk 4 above)
Comment (clouless) [inline war/src/main/webapp/css/style.css]: ok. will do the change in just a sec.

File: war/src/main/webapp/css/style.css (hunk 5 above)
Comment (clouless) [inline war/src/main/webapp/css/style.css]: ok I removed the spaces. but someone should clean that up in a separate pr maybe...

File: war/src/main/webapp/css/style.css (hunk 4 above)
Comment (clouless) [inline war/src/main/webapp/css/style.css]: @recena done.

File: core/src/main/resources/lib/hudson/rssBar.jelly (hunk 3 above)
Comment (recena) [inline core/src/main/resources/lib/hudson/rssBar.jelly]: @clouless `id` are very helpful when you are identifying high-level UI components or containers, specially, in the layout. And you know for sure that you won't have the same component in the UI twice (since the `id` must be unique).

File: core/src/main/resources/lib/hudson/rssBar.jelly [hunk 6]
```diff
@@ -24,22 +24,22 @@ THE SOFTWARE.
...
-      <a href="rssFailed"><img src="${imagesURL}/atom.gif" border="0" alt="Feed" height="16" width="16"/></a>
+    <span id="rss-bar-rss-failed" class="rss-bar-item">
+      <a href="rssFailed" class="rss-bar-item-icon-link"><img src="${imagesURL}/atom.gif" class="icon-rss icon-sm" alt="Feed"/></a>
       <st:nbsp/>
-      <a href="rssFailed">RSS ${%for failures}</a>
+      <a href="rssFailed" class="rss-bar-item-text-link">RSS ${%for failures}</a>
     </span>
-    <span style="padding-left:1em">
-      <a href="rssLatest"><img src="${imagesURL}/atom.gif" border="0" alt="Feed" height="16" width="16"/></a>
+    <span id="rss-bar-rss-latest" class="rss-bar-item">
```
Comment (recena) [inline core/src/main/resources/lib/hudson/rssBar.jelly]: `rss-bar-rss-latest` why? When you can identify this element by using CSS Pseudo-classes.

Comment (clouless) [inline core/src/main/resources/lib/hudson/rssBar.jelly]: If the code would look like this:

```html
    <span class="rss-bar-item">
      <a href="rssAll" class="rss-bar-item-icon-link"><img src="${imagesURL}/atom.gif" class="icon-rss icon-sm" alt="Feed"/></a>
      <st:nbsp/>
      <a href="rssAll" class="rss-bar-item-text-link">RSS ${%for all}</a>
    </span>
    <span class="rss-bar-item">
      <a href="rssFailed" class="rss-bar-item-icon-link"><img src="${imagesURL}/atom.gif" class="icon-rss icon-sm" alt="Feed"/></a>
      <st:nbsp/>
      <a href="rssFailed" class="rss-bar-item-text-link">RSS ${%for failures}</a>
    </span>
    <span class="rss-bar-item">
      <a href="rssLatest" class="rss-bar-item-icon-link" ><img src="${imagesURL}/atom.gif" class="icon-rss icon-sm" alt="Feed"/></a>
      <st:nbsp/>
      <a href="rssLatest" class="rss-bar-item-text-link">RSS ${%for just latest builds}</a>
    </span>
```


How would I identify the img of the rssFailed?
Are you suggesting to use CSS:

```css
#rss-bar-item a[href$=rssFailed] img {
   border:2px solid red;
}
```

Because then the UI breaks if someone changes the href to something else.
That's why I like to have ids, but if you are persistent, I can remove the IDs and
use - the in my optinion ugly - pseudo selectors in my external css.

File: core/src/main/resources/lib/hudson/rssBar.jelly (hunk 3 above)
Comment (clouless) [inline core/src/main/resources/lib/hudson/rssBar.jelly]: @recena I know. And usually when writing other software I don't overuse IDs too much as well. Please see my other response below.

**Comment by clouless** [comment]
@recena just a small reminder :) Maybe you have a lot to do, but I think now it is ready to merge :)

**Comment by oleg-nenashev** [comment]
@clouless There was Jenkins World this week, so all reviews may be delayed.

**Review by oleg-nenashev — APPROVED** [review]
Looks good

**Review by daniel-beck — APPROVED** [review]

File: war/src/main/webapp/css/style.css (hunk 4 above)
Comment (recena) [reply war/src/main/webapp/css/style.css]: Finally this change request was not addressed.

**Comment by recena** [comment]
This was merged with a change request without addressing.

**Comment by daniel-beck** [comment]
@recena Please indicate blocking requests by a negative code review ("requesting changes").

**Comment by oleg-nenashev** [comment]
Yes. We were reviewing pull-requests together with Daniel at JW, and this change made a full impression of a ready pull-request with addressed comments. Even if there are changes required, they can be delivered in the follow-up PR

File: war/src/main/webapp/css/style.css (hunk 4 above)
Comment (clouless) [reply war/src/main/webapp/css/style.css]: @recena it was addressed. Please look at the final code.

**Comment by clouless** [comment]
Are there any followups? I don't see any. If yes, please state them clearly, so I can address them, thx :)

**Comment by recena** [comment]
@daniel-beck @oleg-nenashev My mistake.

@clouless In `master` I see [that](https://github.com/jenkinsci/jenkins/blob/master/war/src/main/webapp/css/style.css#L1919).

**Comment by csgruenebe** [comment]
ok and that is what you wanted me to do?! So what exactly is not right?

**Comment by clouless** [comment]
ok and that is what you wanted me to do?! So what exactly is not right?

**Comment by recena** [comment]
@clouless Yes, to follow the conding style convention.

**Comment by clouless** [comment]
ok so you are talking about the white spaces inside css? If yes I can fix that if you want. But we would be faster, if you clearly state what you want me to do :)

**Comment by recena** [comment]
@clouless I'll do it 😄

**Comment by clouless** [comment]
there you go :) 
#3003

**Comment by recena** [comment]
@clouless I had been already included the changes in [PR](https://github.com/jenkinsci/jenkins/pull/2857/files#diff-62a1ef94d926b7d7d805bae25da89cfbR2025).

**Comment by clouless** [comment]
ok cool :) In the next PR regarding classes/ids I now know what to watch out for :)
